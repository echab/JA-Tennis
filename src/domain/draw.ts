import { Player } from "./player";
import { RankString, ScoreString } from "./types";

export const enum DrawType {
  Knockout, // = 0
  Final, // = 1
  Roundrobin, // = 2
  RoundrobinReturn, // = 3
}
export const enum Mode {
  Build,
  Plan,
  Play,
  Lock,
}

export interface Draw {
  id: string; //new draw has no id

  name: string;

  type: DrawType; //Normal, Final, Poule simple, Poule aller/retour

  suite?: boolean;

  minRank: RankString;
  maxRank: RankString;

  //nbEntry: number;
  nbColumn: number;
  nbOut: number;

  orientation?: number; //Default, Portrait, Landscape

  //matches: Match[];
  boxes: (PlayerIn | Match)[];

  mode?: Mode;

  //_refresh?: Date; //force refresh //TODO?
}

//export interface IDrawDimensions {
//    boxWidth: number;
//    boxHeight: number;
//    interBoxWidth: number;
//    interBoxHeight: number;
//}

export interface Match extends Box {
  //winner: number; //1 or 2 (or undefined)
  score: ScoreString; //a match is a box with a score member
  wo?: boolean;
  qualifOut?: number;

  canceled?: boolean;
  vainqDef?: boolean; //TODO english

  //Planning
  place?: number;
  date?: Date;

  matchFormat?: string; //FFT extent

  note?: string;

  _player1?: Player; //TODO for planning and dialog match
  _player2?: Player;
}

export interface PlayerIn extends Box {
  order?: number;
  qualifIn?: number;
  seeded?: number;
}

export interface Box {
  // id: string;
  position: number;

  hidden?: boolean;
  locked?: boolean;

  playerId?: string;

  //Planning
  receive?: boolean;
  aware?: boolean;
}
