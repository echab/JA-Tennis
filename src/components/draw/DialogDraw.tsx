import { Component, For, JSX, onMount, onCleanup } from 'solid-js';
import { Draw, DrawType } from '../../domain/draw';
import { drawLib, GenerateType } from '../../services/draw/drawLib';
import { OptionalId } from '../../domain/object';
import { Player } from '../../domain/player';
import { TEvent } from '../../domain/tournament';
import { RankString, CategoryString } from '../../domain/types';
import { deleteDraw, groupDraw, groupFindAllPlayerOut } from '../../services/drawService';
import { getRegisteredPlayers, ranksName } from '../../services/tournamentService';
import { rank, category } from '../../services/types';
import { columnMax, countInCol } from '../../utils/drawUtil';
import { useForm } from '../util/useForm';
import { commandManager } from '../../services/util/commandManager';
import { selectDraw } from '../util/selection';
import { IconSexe } from '../misc/IconSexe';

const EMPTY: OptionalId<Draw> = { name: '', type: DrawType.Knockout, minRank: '', maxRank: '', nbColumn: 3, nbOut: 1, boxes: [] };

type Props = {
  event: TEvent;
  draw?: OptionalId<Draw>;
  allPlayers: Player[];
  onOk: (event: TEvent, draws: OptionalId<Draw>[]) => void;
  onClose: () => void;
}

export const DialogDraw: Component<Props> = (props) => {
  let refDlg: HTMLDialogElement;

  onMount(() => {
    refDlg.addEventListener('close', props.onClose)
    refDlg.showModal();
  });

  onCleanup(() => {
    refDlg.removeEventListener('close', props.onClose)
  })

  const draw: OptionalId<Draw> | undefined = props.draw && { ...props.draw }; // clone, without reactivity

  const isFirstDraw = !props.event.draws.length || draw?.id === props.event.draws[0].id;

  const { form, setForm, updateField } = useForm<OptionalId<Draw>>(draw ?? EMPTY);

  const ranks: RankString[] = rank.list();

  if (!props.draw) {
    // new draw
    const prevRank = props.event.draws.at(-1)?.maxRank;
    const minRank = prevRank ? rank.next(prevRank) : rank.first();
    setForm({ minRank, maxRank: props.event.maxRank });
  }

  const drawTypes: { value: number; label: string; }[] = [];
  drawTypes[DrawType.Knockout] = { value: DrawType.Knockout, label: "Knock out" };
  drawTypes[DrawType.Final] = { value: DrawType.Final, label: "Final" };
  drawTypes[DrawType.Roundrobin] = { value: DrawType.Roundrobin, label: "Round-robin" };
  drawTypes[DrawType.RoundrobinReturn] = { value: DrawType.RoundrobinReturn, label: "Round-robin with return" };

  const categories: CategoryString[] = category.list();

  /** @returns registered players and entries numbers from previous draw */
  const registeredPlayersOrQ = (): (Player|number)[] => {
    const players: (Player|number)[] = getRegisteredPlayers(props.allPlayers, props.event, form.minRank, form.maxRank);
    const prevDraw = props.event.draws.at(-1) as Draw | undefined;
    if (prevDraw) {
      const [iStart, iNext] = groupDraw(props.event, prevDraw);
      if (iStart < iNext) {
          const qualifs = groupFindAllPlayerOut(props.event, [iStart, iNext]);
          if (qualifs.length) {
            return players.concat(qualifs);
          }
      }
    }
    return players;
  }

  const getNbEntry = (): number => {
    return form.type & DrawType.Roundrobin
      ? form.nbColumn // TODO type poule
      : countInCol(columnMax(form.nbColumn, form.nbOut), form.nbOut);
  }

  const submit: JSX.EventHandlerUnion<HTMLFormElement, Event & { submitter: HTMLElement; }> = (evt) => {
    evt.preventDefault();

    // const formElems = (evt.target as HTMLFormElement).elements as unknown as Record<keyof Draw, RadioNodeList>;

    let result: OptionalId<Draw>[] = [{
      id: form.id || undefined,
      name: form.name.trim() || ranksName(form.minRank, form.maxRank),
      type: form.type,
      suite: form.suite,

      minRank: form.minRank,
      maxRank: form.maxRank,

      nbColumn: form.nbColumn,
      nbOut: form.nbOut,

      orientation: form.orientation,

      boxes: draw?.boxes ?? [],

      mode: form.mode,
    }];

    if ((evt.submitter as HTMLButtonElement).value === 'generate') {
      const lib = drawLib(props.event, result[0] as Draw);
      result = lib.generateDraw(GenerateType.Create, registeredPlayersOrQ());
    }

    props.onOk(props.event, result);
    
    refDlg.close();
    // props.onClose();
  };

  const deleteAndClose = () => {
    if (draw?.id) {
      commandManager.add(deleteDraw(draw.id));
      selectDraw(props.event, undefined);
      refDlg.close();
      // props.onClose();
    }
  }

  return (
    <dialog ref={refDlg!} class="p-0">
      <header class="flex justify-between sticky top-0 bg-slate-300 p-1">
        <span><i class='icon2-draw'></i> <IconSexe sexe={props.event.sexe} double={props.event.typeDouble} />{props.event.name} - <b>{props.draw ? `Edit draw ${props.draw?.name ?? ''}` : 'New draw'}</b></span>
        <small>Id: {props.draw?.id}</small>
        <button type="button" data-dismiss="modal" aria-hidden="true"
          onclick={() => refDlg.close()}
        >&times;</button>
      </header>
      <form method="dialog" class="w-[32rem]" onsubmit={submit}>
        <div class="p-4">
          <input id="id" type="hidden" value={form.id} />
          <div class="mb-1">
            <label for="name" class="inline-block w-3/12 text-right pr-3">Name:</label>
            <input id="name" type="text" class="w-9/12 p-1"
              autofocus
              placeholder={ranksName(form.minRank, form.maxRank)}
              value={form.name !== ranksName(form.minRank, form.maxRank) ? form.name : ''} onChange={updateField("name")} />
            {/*<span class="error" show.bind="eventForm.name.$error.required">Required!</span> */}
          </div>

          <div class="mb-1">
            <label for="type" class="inline-block w-3/12 text-right pr-3">Type:</label>
            <select id="type" required value={form.type} onChange={updateField('type')} class="w-6/12 p-1">
              <For each={drawTypes}>{({ value, label }) => <option value={value}>{label}</option>}</For>
            </select>
          </div>

          <div class="mb-1">
            <label for="minRank" class="inline-block w-3/12 text-right pr-3">Rank, from:</label>
            <select id="minRank" value={form.minRank} onChange={updateField('minRank')} class="w-2/12 p-1">
              <For each={ranks}>{(rank) => <option>{rank}</option>}</For>
            </select>

            <label for="maxRank" class="inline-block w-2/12 text-right pr-3">to:</label>
            <select id="maxRank" value={form.maxRank} onChange={updateField('maxRank')} class="w-2/12 p-1">
              <For each={ranks}>{(rank) => <option>{rank}</option>}</For>
            </select>
            <label class="pl-3">{registeredPlayersOrQ().length} registered</label>
          </div>

          <div class="mb-1">
            <span class="inline-block w-3/12 text-right pr-3"></span>
            <label><input type="checkbox" checked={form.suite} onchange={updateField('suite')}
              disabled={isFirstDraw}
              /> Same group as previous draw</label>
          </div>

          <fieldset>
            <legend>Dimensions:</legend>
            <div class="mb-1">
              <label for="nbIn" class="inline-block w-3/12 text-right pr-3">Entries:</label>
              <input id="nbIn" type="number" readonly value={getNbEntry()} class="w-3/12 p-1"/>
            </div>
            <div class="mb-1">
              <label for="nbColumn" class="inline-block w-3/12 text-right pr-3">Columns:</label>
              <input id="nbColumn" type="number" value={form.nbColumn} onChange={updateField('nbColumn')} min="2" max="9" required class="w-3/12 p-1" />
            </div>
            <div class="mb-1">
              <label for="nbOut" class="inline-block w-3/12 text-right pr-3">Outputs:</label>
              <input id="nbOut" type="number" value={form.nbOut} onChange={updateField('nbOut')} min="1" max="16" required class="w-3/12 p-1" />
            </div>
          </fieldset>

          {/* 
          TODO paper orientation
          */}

        </div>
        <footer class='sticky bottom-0 flex justify-end space-x-2 mt-2 pb-4 pr-4 bg-gray-50 bg-opacity-60'>
          <button type="submit" value="generate"
            //  disabled={!!drawForm.$error.required}
             class="rounded-md border border-transparent bg-indigo-200 py-2 px-4 min-w-[6rem]">
              <i class="icon2-draw-generate"></i> Generate
          </button>
          <button type="submit"
            //  disabled.bind="!!eventForm.$error.required"
            class="rounded-md border border-transparent bg-indigo-400 py-2 px-4 min-w-[6rem]"
          >OK
          </button>

          <button type="button" class="rounded-md border border-transparent bg-gray-200 py-2 px-4 min-w-[6rem]"
            value="Delete" disabled={!form.id}
            onclick={deleteAndClose}
            >âœ– Delete
          </button>

          <button type="button" class="rounded-md border border-transparent bg-gray-200 py-2 px-4 min-w-[6rem]"
            data-dismiss="modal" aria-hidden="true"
            onclick={() => refDlg.close()}
          >Cancel</button>
        </footer>
        {/*{ eventForm.$error } */}
      </form>
    </dialog>
  );
};
