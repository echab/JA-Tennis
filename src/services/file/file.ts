import { Tournament } from "../../domain/tournament";
import { convert, docFields, jatFileType } from "./jatSchema";
import { createSerializer } from "./serializer";

let fileName: string = '';

export async function openFile(): Promise<Tournament> {
    // alert('hello');
    const w: any = window;
    const [fileHandle] = await w.showOpenFilePicker({
        types: [jatFileType],
        excludeAcceptAllOption: true,
        multiple: false,
    });
    return readFile(fileHandle);
}

async function readFile(fileHandle: any): Promise<Tournament> {
    console.log('reading ', fileHandle.name);
    const file = await fileHandle.getFile();
    const { size, lastModifiedDate, type } = file;
    console.log('size=', size, 'lastModifiedDate=', lastModifiedDate);
    // const content = await file.stream();
    // for await (const value of streamAsyncIterator(content)) {
    const buffer = await file.arrayBuffer();
    const value = new Uint8Array(buffer);

    const reader = createSerializer(true, value, 0);
    const doc = reader.deserialize(docFields, NaN, 'doc');
    console.log(doc);

    fileName = fileHandle.name;

    return convert(doc);
}

export async function saveFile() {
    const w: any = window;
    const fileHandle = await w.showSaveFilePicker({
        types: [jatFileType],
        suggestedName: fileName
    });
    const writable = await fileHandle.createWritable();
    // await writable.write(contents);
    await writable.close();
}

function streamAsyncIterator(stream: ReadableStream, flags?: { mode: "byob" }) {
    // Get a lock on the stream:
    const reader = stream.getReader(); // stream.getReader(flags);

    return {
        next() {
            // Stream reads already resolve with {done, value}, so
            // we can just call read:
            return reader.read();
        },
        return() {
            // Release the lock if the iterator terminates.
            reader.releaseLock();
            return {};
        },
        // for-await calls this on whatever it's passed, so
        // iterators tend to return themselves.
        [Symbol.asyncIterator]() {
            return this;
        }
    };
}

/* TODO
    document.getElementById('btnOpen').addEventListener('click', openFile);
    document.getElementById('btnSave').addEventListener('click', saveFile);
    const dropZone = document.getElementById('dropZone');
    dropZone.addEventListener('dragover', (event) => {
        event.preventDefault();
        dropZone.style.backgroundColor = '#ddd';
    });
    dropZone.addEventListener('drop', async (event) => {
        event.preventDefault();
        const fileHandlesPromises = [...event.dataTransfer.items]
            .filter((item) => item.kind === 'file')
            .map((item: any) => item.getAsFileSystemHandle());

        for await (const handle of fileHandlesPromises) {
            if (handle.kind === 'directory') {
                console.log(`Directory: ${handle.name}`);
            } else {
                console.log(`File: ${handle.name}`);
                readFile(handle);
            }
        }
    });
*/