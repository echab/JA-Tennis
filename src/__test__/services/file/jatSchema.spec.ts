/// <reference types="node" />
import { DrawType } from "../../../domain/draw";
import type { Tournament } from "../../../domain/tournament";
import { docFields } from "../../../services/file/jatSchema";
import { createSerializer } from "../../../services/file/serializer";
import { readFile } from 'node:fs/promises';

describe("jatSchema", () => {
    const tournament2: Tournament = {
        version: 13,
        id: 'T1',
        info: { name: 'Tournament 1', slotLength: 90 },
        types: { name: 'FFT', versionTypes: 5 },
        players: [
            { id: 'J1', name: 'Albert', sexe: 'H', rank: 'NC', registration: ['E1'] },
            { id: 'J2', name: 'Bertty', sexe: 'H', rank: '40', registration: ['E1'] },
            { id: 'J3', name: 'Ernest', sexe: 'H', rank: '30/5', registration: ['E1'] },
            { id: 'J4', name: 'Freddy', sexe: 'H', rank: '30/4', registration: [] },
        ],
        events: [
            {
                id: 'E1', name: 'Simple mens', sexe: 'H', category: 11, maxRank: '30/1', draws: [
                    {
                        id: 'D11', name: 'First draw', type: DrawType.Knockout, nbColumn: 2, nbOut: 2, minRank: 'NC', maxRank: '40', boxes: [
                            { position: 6, playerId: 'J1' },
                            { position: 2, score: '6/4 6/1', date: new Date('2022-11-03T20:30Z'), playerId: 'J1', qualifOut: 1 },
                            { position: 5, playerId: 'J2' },
                            { position: 4, playerId: 'J3' },
                            { position: 1, score: '', place: 0, qualifOut: 2 },
                            { position: 3, playerId: 'J4' },
                        ]
                    }
                ]
            }
        ]
    };

    it('should read binary tournament1.jat file version 13', async () => {
        const b = await readFile(`${__dirname}/tournament1.jat`);
        const buf = b.buffer;

        const reader = createSerializer(new Uint8Array(buf));

        const result = reader.readObject(docFields);

        expect(result.version).toBe(13);
        expect(result).toMatchSnapshot();
    });

    it('should save tournament to array buffer', () => {
        const writer = createSerializer(new Uint8Array(8192)); // TODO size
        writer.writeObject(tournament2, docFields, docFields.version.def);

        expect(writer._buffer).toMatchSnapshot();
    });
});
