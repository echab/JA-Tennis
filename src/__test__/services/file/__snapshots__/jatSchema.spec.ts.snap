// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jatSchema should read binary tournament1.jat file version 13 1`] = `
{
  "_nDay": 4,
  "events": [
    {
      "category": 7,
      "color": "#ffffff",
      "consolation": 0,
      "dateMaj": 2022-11-03T11:00:00.000Z,
      "draws": [
        {
          "boxes": [
            {
              "date": 2022-11-03T10:00:00.000Z,
              "fm": 0,
              "order": 0,
              "place": 0,
              "position": 1,
              "qualifOut": 2,
              "score": "",
              "version": 4,
            },
            {
              "date": 2022-11-03T14:00:00.000Z,
              "fm": 0,
              "order": -1,
              "place": 1,
              "playerId": "1",
              "position": 2,
              "qualifOut": 1,
              "score": "6/4 6/2",
              "version": 4,
            },
            {
              "fm": 0,
              "order": 4,
              "playerId": "3",
              "position": 3,
              "seeded": 1,
              "version": 4,
            },
            {
              "fm": 0,
              "order": 3,
              "playerId": "8",
              "position": 4,
              "version": 4,
            },
            {
              "fm": 0,
              "order": 2,
              "playerId": "0",
              "position": 5,
              "version": 4,
            },
            {
              "fm": 0,
              "order": 1,
              "playerId": "1",
              "position": 6,
              "seeded": 2,
              "version": 4,
            },
          ],
          "dateMaj": 2022-11-03T11:00:00.000Z,
          "formatMatch": 0,
          "id": "885",
          "maxRank": "30/3",
          "minRank": "NC",
          "name": "NC - 30/1",
          "nbColumn": 2,
          "nbOut": 2,
          "orientation": 1,
          "suite": 0,
          "type": 0,
          "version": 10,
        },
        {
          "boxes": [
            {
              "fm": 0,
              "order": 0,
              "position": 0,
              "score": "",
              "version": 4,
            },
            {
              "fm": 0,
              "order": 0,
              "position": 1,
              "score": "",
              "version": 4,
            },
            {
              "fm": 0,
              "order": 0,
              "position": 2,
              "score": "",
              "version": 4,
            },
            {
              "date": 2022-11-04T19:15:00.000Z,
              "fm": 0,
              "order": 0,
              "place": 0,
              "position": 3,
              "score": "",
              "version": 4,
            },
            {
              "fm": 0,
              "note": "WO",
              "order": -9,
              "playerId": "5",
              "position": 4,
              "score": "WO",
              "version": 4,
            },
            {
              "fm": 0,
              "note": "Un VDef
Jean-Charle is receiving.",
              "order": -7,
              "playerId": "2",
              "position": 5,
              "receive": true,
              "score": "6/0 6/0 VD",
              "version": 4,
            },
            {
              "date": 2022-11-03T09:00:00.000Z,
              "fm": 0,
              "order": 0,
              "position": 6,
              "score": "",
              "version": 4,
            },
            {
              "fm": 0,
              "order": 10,
              "playerId": "9",
              "position": 7,
              "seeded": 1,
              "version": 4,
            },
            {
              "fm": 0,
              "order": 0,
              "position": 8,
              "qualifIn": 0,
              "version": 4,
            },
            {
              "fm": 0,
              "order": 9,
              "playerId": "5",
              "position": 9,
              "version": 4,
            },
            {
              "fm": 0,
              "order": 8,
              "playerId": "4",
              "position": 10,
              "version": 4,
            },
            {
              "fm": 0,
              "order": 7,
              "playerId": "2",
              "position": 11,
              "version": 4,
            },
            {
              "fm": 0,
              "order": 6,
              "playerId": "6",
              "position": 12,
              "version": 4,
            },
            {
              "fm": 0,
              "order": -1,
              "playerId": "1",
              "position": 13,
              "qualifIn": 1,
              "version": 4,
            },
            {
              "fm": 0,
              "order": 5,
              "playerId": "7",
              "position": 14,
              "seeded": 2,
              "version": 4,
            },
          ],
          "dateMaj": 2022-11-03T11:00:00.000Z,
          "formatMatch": 0,
          "id": "19450",
          "lock": 1,
          "maxRank": "15",
          "minRank": "30/2",
          "name": "Tableau Final",
          "nbColumn": 4,
          "nbOut": 1,
          "orientation": 1,
          "suite": 0,
          "type": 1,
          "version": 10,
        },
      ],
      "formatMatch": 0,
      "id": "1191",
      "maxRank": "15/3",
      "name": "Simple Mens",
      "sexe": "H",
      "version": 10,
    },
    {
      "category": -1,
      "color": "#ff80ff",
      "consolation": 1,
      "dateMaj": 2022-11-03T11:00:00.000Z,
      "draws": [
        {
          "boxes": [
            {
              "fm": 0,
              "hidden": true,
              "order": 0,
              "position": 0,
              "score": "",
              "version": 4,
            },
            {
              "fm": 0,
              "order": 0,
              "position": 1,
              "score": "",
              "version": 4,
            },
            {
              "fm": 0,
              "order": 0,
              "position": 2,
              "score": "",
              "version": 4,
            },
            {
              "fm": 0,
              "hidden": true,
              "order": 0,
              "position": 3,
              "score": "",
              "version": 4,
            },
            {
              "fm": 0,
              "hidden": true,
              "order": 0,
              "position": 4,
              "score": "",
              "version": 4,
            },
            {
              "fm": 0,
              "order": -1,
              "playerId": "10",
              "position": 5,
              "score": "6/1 6/7 6/0",
              "version": 4,
            },
            {
              "fm": 0,
              "hidden": true,
              "order": 0,
              "position": 6,
              "score": "",
              "version": 4,
            },
            {
              "fm": 0,
              "hidden": true,
              "order": 0,
              "position": 7,
              "score": "",
              "version": 4,
            },
            {
              "fm": 0,
              "hidden": true,
              "order": 0,
              "position": 8,
              "score": "",
              "version": 4,
            },
            {
              "fm": 0,
              "order": 0,
              "position": 9,
              "version": 4,
            },
            {
              "fm": 0,
              "order": 2,
              "playerId": "13",
              "position": 10,
              "version": 4,
            },
            {
              "fm": 0,
              "order": 1,
              "playerId": "10",
              "position": 11,
              "seeded": 1,
              "version": 4,
            },
          ],
          "dateMaj": 2022-11-03T11:00:00.000Z,
          "formatMatch": 0,
          "id": "6723",
          "lock": 1,
          "maxRank": "30/1",
          "minRank": "NC",
          "name": "Roundrobin1",
          "nbColumn": 3,
          "nbOut": 1,
          "orientation": 0,
          "suite": 0,
          "type": 2,
          "version": 10,
        },
        {
          "boxes": [
            {
              "fm": 0,
              "hidden": true,
              "order": 0,
              "position": 0,
              "score": "",
              "version": 4,
            },
            {
              "fm": 0,
              "order": 0,
              "position": 1,
              "score": "",
              "version": 4,
            },
            {
              "fm": 0,
              "order": 0,
              "position": 2,
              "score": "",
              "version": 4,
            },
            {
              "fm": 0,
              "order": 0,
              "position": 3,
              "score": "",
              "version": 4,
            },
            {
              "fm": 0,
              "hidden": true,
              "order": 0,
              "position": 4,
              "score": "",
              "version": 4,
            },
            {
              "fm": 0,
              "order": 0,
              "position": 5,
              "score": "",
              "version": 4,
            },
            {
              "fm": 0,
              "order": 0,
              "position": 6,
              "score": "",
              "version": 4,
            },
            {
              "fm": 0,
              "order": 0,
              "position": 7,
              "score": "",
              "version": 4,
            },
            {
              "fm": 0,
              "hidden": true,
              "order": 0,
              "position": 8,
              "score": "",
              "version": 4,
            },
            {
              "fm": 0,
              "order": 0,
              "position": 9,
              "version": 4,
            },
            {
              "fm": 0,
              "order": 0,
              "position": 10,
              "version": 4,
            },
            {
              "fm": 0,
              "order": 0,
              "position": 11,
              "version": 4,
            },
          ],
          "dateMaj": 2022-11-03T11:00:00.000Z,
          "formatMatch": 0,
          "id": "12039",
          "maxRank": "30/1",
          "minRank": "NC",
          "name": "NC - 30/1",
          "nbColumn": 3,
          "nbOut": 1,
          "orientation": 0,
          "suite": 1,
          "type": 3,
          "version": 10,
        },
      ],
      "formatMatch": 0,
      "id": "1827",
      "maxRank": "30/1",
      "name": "Girls consolation",
      "sexe": "F",
      "version": 10,
    },
    {
      "category": 11,
      "color": "#ffbbd5",
      "consolation": 0,
      "dateMaj": 2022-11-03T11:00:00.000Z,
      "draws": [
        {
          "boxes": [
            {
              "fm": 2,
              "order": 0,
              "position": 0,
              "score": "",
              "version": 4,
            },
            {
              "fm": 2,
              "order": 2,
              "playerId": "14",
              "position": 1,
              "version": 4,
            },
            {
              "fm": 2,
              "order": 1,
              "playerId": "12",
              "position": 2,
              "version": 4,
            },
          ],
          "dateMaj": 2022-11-03T11:00:00.000Z,
          "formatMatch": 0,
          "id": "10311",
          "maxRank": "30/1",
          "minRank": "NC",
          "name": "NC - 30/1",
          "nbColumn": 2,
          "nbOut": 1,
          "orientation": 0,
          "suite": 0,
          "type": 0,
          "version": 10,
        },
      ],
      "end": 2022-11-17T11:00:00.000Z,
      "formatMatch": 2,
      "id": "3787",
      "maxRank": "30/1",
      "name": "Double mixte",
      "sexe": "M",
      "start": 2022-11-03T11:00:00.000Z,
      "typeDouble": true,
      "version": 10,
    },
  ],
  "id": "24250",
  "info": {
    "club": {
      "adress1": "rue de la Gare",
      "adress2": "31120 Roques-sur-Garonne",
      "ligue": "Midi-Pyrénées",
      "name": "TROC",
      "version": 13,
    },
    "end": 2022-11-06T11:00:00.000Z,
    "homologation": "1234567890",
    "name": "Jeu de test",
    "referee": {
      "adress1": "Lotissement du Verger Sud",
      "adress2": "Portet",
      "name": "Geneviève",
      "version": 13,
    },
    "slotLength": 90,
    "start": 2022-11-03T11:00:00.000Z,
    "version": 13,
  },
  "places": [
    {
      "avail": [
        -1,
        -1,
        -1,
        -1,
      ],
      "name": "Court 1",
      "version": 13,
    },
    {
      "avail": [
        -1056968449,
        -1,
        -1,
        -1,
      ],
      "name": "Salle B",
      "version": 13,
    },
  ],
  "players": [
    {
      "avail": [
        4294967295,
        4294967295,
        4294967295,
        4294967295,
      ],
      "category": undefined,
      "firstname": "Gérard",
      "id": "0",
      "licence": 4310365,
      "name": "BERANDON",
      "rank": "NC",
      "registration": [
        "1191",
      ],
      "sexe": "H",
      "version": 10,
    },
    {
      "avail": [
        4294967295,
        4294967295,
        4294967295,
        4294967295,
      ],
      "category": undefined,
      "firstname": "Denis",
      "id": "1",
      "licence": 3464539,
      "name": "ARGENTON",
      "rank": "NC",
      "registration": [
        "1191",
      ],
      "sexe": "H",
      "version": 10,
    },
    {
      "avail": [
        4294967295,
        4294967295,
        4294967295,
        4294967295,
      ],
      "category": undefined,
      "dateMaj": 2022-11-03T11:00:00.000Z,
      "firstname": "Jean Charles",
      "id": "2",
      "licence": 3720420,
      "name": "D'HUY",
      "rank": "30/2",
      "rank2": "30/2",
      "registration": [
        "1191",
      ],
      "sexe": "H",
      "version": 10,
    },
    {
      "avail": [
        4294967295,
        4294967295,
        4294967295,
        4294967295,
      ],
      "category": undefined,
      "dateMaj": 2022-11-03T11:00:00.000Z,
      "firstname": "Serge",
      "id": "3",
      "licence": 2648972,
      "name": "DAMPERT",
      "rank": "30/3",
      "rank2": "30/3",
      "registration": [
        "1191",
      ],
      "sexe": "H",
      "version": 10,
    },
    {
      "adress1": "1 rue des Fleurs",
      "adress2": "Au bout de la rue",
      "avail": [
        4294967295,
        4294967295,
        4294967295,
        4294967295,
      ],
      "birth": 1971-05-24T11:00:00.000Z,
      "category": "categ51",
      "city": "ROQUES SUR GARONNE",
      "club": "TROC",
      "comment": "Tous les jours
Soir et weekend.",
      "dateMaj": 2022-11-03T11:00:00.000Z,
      "email": "bruno.angibot@wanadoo.fr",
      "firstname": "Bruno",
      "id": "4",
      "licence": 3724346,
      "name": "ANGIBOT",
      "nationality": "F",
      "phone1": "05 61 78 15 12",
      "phone2": "06 14 45 78 99",
      "rank": "30/1",
      "rank2": "30/1",
      "registration": [
        "1191",
      ],
      "sexe": "H",
      "solde": 2500,
      "soldeCheque": 1000,
      "soldeEspece": 500,
      "soldeType": 1,
      "version": 10,
      "zipCode": "31120",
    },
    {
      "avail": [
        4294967295,
        4294967295,
        4294967295,
        4294967295,
      ],
      "category": undefined,
      "dateMaj": 2022-11-03T11:00:00.000Z,
      "firstname": "Matthieu",
      "id": "5",
      "licence": 1293078,
      "name": "CARMET",
      "rank": "30",
      "rank2": "30",
      "registration": [
        "1191",
      ],
      "sexe": "H",
      "version": 10,
    },
    {
      "avail": [
        4294967295,
        4294967295,
        4294967295,
        4294967295,
      ],
      "category": undefined,
      "dateMaj": 2022-11-03T11:00:00.000Z,
      "firstname": "Jérôme",
      "id": "6",
      "licence": 1294209,
      "name": "ASTRUC",
      "rank": "15/5",
      "rank2": "15/5",
      "registration": [
        "1191",
      ],
      "sexe": "H",
      "version": 10,
    },
    {
      "avail": [
        4294967295,
        4294967295,
        4294967295,
        4294967295,
      ],
      "category": undefined,
      "dateMaj": 2022-11-03T11:00:00.000Z,
      "firstname": "Philippe",
      "id": "7",
      "licence": 194539,
      "name": "BRUN",
      "rank": "15/3",
      "rank2": "15/3",
      "registration": [
        "1191",
      ],
      "sexe": "H",
      "version": 10,
    },
    {
      "avail": [
        4294967295,
        4294967295,
        4294967295,
        4294967295,
      ],
      "category": undefined,
      "dateMaj": 2022-11-03T11:00:00.000Z,
      "firstname": "Hans",
      "id": "8",
      "licence": 5174622,
      "name": "BAYERSDORFF",
      "phone1": "05.01.02.03.04",
      "phone2": "06.02.04.08.10",
      "registration": [],
      "sexe": "H",
      "version": 10,
    },
    {
      "avail": [
        4294967295,
        4294967295,
        4294967295,
        4294967295,
      ],
      "category": undefined,
      "dateMaj": 2022-11-03T11:00:00.000Z,
      "firstname": "Jean-Paul",
      "id": "9",
      "licence": 490377,
      "name": "BOURDEAU",
      "rank": "15/2",
      "rank2": "15/2",
      "registration": [
        "1191",
      ],
      "sexe": "H",
      "version": 10,
    },
    {
      "avail": [
        4294967295,
        4294967295,
        4294967295,
        4294967295,
      ],
      "birth": "2007",
      "category": "categ15",
      "dateMaj": 2022-11-03T11:00:00.000Z,
      "firstname": "Marie",
      "id": "10",
      "name": "POPINS",
      "nationality": "FRANÇAISE",
      "rank": "-1/6",
      "registration": [],
      "sexe": "F",
      "version": 10,
    },
    {
      "avail": [
        4294967295,
        4294967295,
        4294967295,
        4294967295,
      ],
      "category": undefined,
      "dateMaj": 2022-11-03T11:00:00.000Z,
      "id": "11",
      "name": "ANGIBOT B - ARGENTON D",
      "registration": [],
      "sexe": "H",
      "teamIds": [
        "4",
        "1",
      ],
      "version": 10,
    },
    {
      "avail": [
        4294967295,
        4294967295,
        4294967295,
        4294967295,
      ],
      "category": undefined,
      "dateMaj": 2022-11-03T11:00:00.000Z,
      "id": "12",
      "name": "ASTRUC J - POPINS M",
      "registration": [
        "3787",
      ],
      "sexe": "M",
      "teamIds": [
        "6",
        "10",
      ],
      "version": 10,
    },
    {
      "avail": [
        4294967295,
        4294967295,
        4294967295,
        4294967295,
      ],
      "category": undefined,
      "dateMaj": 2022-11-03T11:00:00.000Z,
      "firstname": "Anne",
      "id": "13",
      "name": "DUFOUR",
      "registration": [
        "1827",
      ],
      "sexe": "F",
      "version": 10,
    },
    {
      "avail": [
        4294967295,
        4294967295,
        4294967295,
        4294967295,
      ],
      "category": undefined,
      "dateMaj": 2022-11-03T11:00:00.000Z,
      "id": "14",
      "name": "DUFOUR A",
      "registration": [
        "3787",
      ],
      "sexe": "M",
      "teamIds": [
        "13",
      ],
      "version": 10,
    },
  ],
  "types": {
    "name": "FFT",
    "version": 13,
    "versionTypes": 5,
  },
  "version": 13,
}
`;
